src/msg/message.cc

```diff
@@ -19,44 +23,31 @@ class SubscriptionInfo {
     std::vector<CRNode*>             sub_list_;
 };
 
-}  // namespace
+class SubscriptionMap {
+   public:
+    bool Subscribe(const channel_id_t channel, CRNode* node);
 
-using channel_id_t = CRMessageBase::channel_id_t;
+    void Unsubscribe(const channel_id_t channel, CRNode* node);
 
-using subscription_map_t = std::unordered_map<channel_id_t, SubscriptionInfo, boost::hash<channel_id_t>>;
+    void Dispatch(const CRMessageBasePtr& message);
 
-static subscription_map_t& GetSubscriptionMap() {
-    static subscription_map_t subscription_map;
-    return subscription_map;
-};
+    cr_timestamp_nsec_t GetLatestDeliveredTime(const channel_id_t channel) const;
 
-static SubscriptionInfo* GetSubscriptionInfo(const channel_id_t channel) {
-    auto& subscription_map  = GetSubscriptionMap();
-    auto  subscription_find = subscription_map.find(channel);
-    if (subscription_find == subscription_map.end()) {
-        return nullptr;
-    }
-    return &subscription_find->second;
-}
+    static SubscriptionMap& GetInstance();
 
-channel_id_t CRMessageBase::GetChannelId() const {
-    return std::make_pair(GetMessageTypeIndex(), GetChannelSubId());
-}
+   private:
+    mutable std::shared_mutex                                                     mtx_;
```

Comment:
I would remove the const instead.
mutable is to poke a hole and should be avoid whenever possible.

Reply:
mutable mutex is a common case, because we don't modify any states but using mutex to protect them.

Reply:
The problem of poking a hole is, why can't we have another hole? What's the standard here.
With const, what we know is, it's not logically const as ref/const pointer can points outside, but at least the storage is const.
But with mutable, the storage itself is no longer const, so what's the point of keeping const part?

Maybe a const pointer is better?

Reply:
With const pointer you can at least memcpy safely.

Reply:
"memcpy safety" means nothing. This is not a [standard layout class](https://en.cppreference.com/w/cpp/named_req/StandardLayoutType).

The standard here is clear: only mutexes can be mutable. The point of keeping `const` is, you cannot modify your states in this function, so we will be sure that is read-only.

Reply:
The point is, how would you effectively communicate this mutex-only exception to others, so that people:
1. Don't start using mutable elsewhere.
2. Can trust you're not using mutable elsewhere in your codebase when witnessing the existence of mutable.

Reply:
ack